src/compiler/program.ts

```diff
@@ -904,7 +904,13 @@ namespace ts {
             let declarationDiagnostics: Diagnostic[] = [];
 
             if (options.noEmit) {
-                return { diagnostics: declarationDiagnostics, sourceMaps: undefined, emittedFiles: undefined, emitSkipped: true };
+                // Allow only emitting of declarations
+                if (options.declaration) {

```

Comment:
@DanielRosenwasser I don't think this overrides `--noEmitOnError`, just `--noEmit`. I do think we might want to have a separate compiler option for this however, as always emitting declarations could have undesired consequences.